src/mem/ruby/protocol/GPU_VIPER-SQC.sm

```diff
@@ -308,22 +331,52 @@ machine(MachineType:SQC, "GPU SQC (L1 I Cache)")
 
   // Transitions
 
+  // if another request arrives for the same cache line that has a pending
+  // load, put it on the wakeup buffer.  This reduced resource contention since
+  // they won't try again every cycle and will instead only try again once woken
+  // up
+  transition(IV, {Fetch}) {
+      st_stallAndWaitRequest;
+  }
+
   // transitions from base
   transition({I, V}, Repl, I) {TagArrayRead, TagArrayWrite} {
     // since we're evicting something, don't bother classifying as hit/miss
     ic_invCache;
   }
 
-  transition(I, Data, V) {TagArrayRead, TagArrayWrite, DataArrayRead} {
+  // if we got a response for a load where the line is in I, then
+  // another request must have come in that replaced the line in question in
+  // the cache.  Thus, complete this request without allocating the line, but
+  // still deallocate TBE and wakeup any dependent addresses.
+  transition(I, Data) {TagArrayRead, TagArrayWrite, DataArrayRead} {
+    // don't profile this as a hit/miss since it's a reponse from L2,
+    // so we already counted it
+    l_loadDoneMiss;
+    wada_wakeUpAllDependentsAddr;
+    d_deallocateTBE;
+    pr_popResponseQueue;
+  }
```

Comment:
This hits on something else that has been bothering me about the current support.  

Note: this example is for the TCP, even though you asked about the SQC (apparently I can't add a comment on the line in question here since it's not a line I modified in this patch).

When a load returns from the L2 (https://github.com/gem5/gem5/pull/540/files#diff-c3f37e59aaafb17f88001c748f5d290395705147dcbfc4a064566431dc31e71cR278), assuming it's not a GLC/SLC load, the load either does a TCC_Ack if the line is valid or evicts something else to make space.  I'm pretty sure this means we'll never get to the transition(I, TCC_Ack...) transition because we'll force someone else out.  I have a local change I'm testing to resolve this, which basically checks if there is a TBE entry but the line is invalid and if so assumes the line got evicted and treats the load like a bypassing load's response.  This seems right to me -- if we evicted the line, we don't want to put it right back in.  But would appreciate any insight or feedback on if this sounds reasonable or is overoptimizing.

Regarding your SQC comment: it looks like I need to update 343 (https://github.com/gem5/gem5/pull/540/files#diff-1bbe1d71ce75a763a3c6ec0ed1ea0945249815209fc73edbb3ddebfc8590c139R343) to include IV?